### ISSUE TITLE
Performance for 100% CPU workload around 40-80% worse compare to native

### ISSUE THREAD

**Issue by paneq** [docker/for-mac#4769]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 68A9C81C-E517-4007-B216-6738A801400A/20200711160019

### Expected behavior

I expect pure CPU overhead to be around 10%

### Actual behavior

Single thread CPU bound workloads are 40-80% slower when run in a docker container

### Information

- It's a reproducible issue on multiple machines
- macOS Version: Mojave 10.14.6 (18G4032)

### Diagnostic logs

I don't understand where I am supposed to take these from. I only see... :

```
Diagnose succeeded
```

### Steps to reproduce the behavior

1. install 2.6.5p114 in any way you like (i.e. using `rbenv`)
2. `time ruby -e "300_000_000.times{1+1}"`

   ```
       10.46 real        10.08 user         0.14 sys
   ```

3. `ruby --version`

   ```
   ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-darwin18]
   ```

4. `docker pull ruby:2.6.5`
5. `time docker run ruby:2.6.5 ruby -e "300_000_000.times{1+1}"`

   ```
       17.86 real         0.05 user         0.13 sys
   ```

6. `time docker run ruby:2.6.5 ruby -e "1+1"` (to check constant overhead)

   ```
        0.86 real         0.04 user         0.12 sys
   ```

7. I run native & docker version multiple times and the results are always very similar. I have other CPU intensive workloads (npm tests for an application) which I verified to run around 40% slower compared to native execution. We are talking about pure CPU overhead without anything mounted.

**Comment by lemire** [comment]
@paneq 

I am sure you will agree that finding one timing difference does not imply that there is a computational overhead. Many things could be happening under the hood.

Would you try the following?

```
time ruby -e "(1..100_000).inject(:*)"
time docker run ruby:2.6.3 ruby -e "(1..100_000).inject(:*)"
```



On my mac, I am getting very nearly the same time (~3s) with a slight advantage for the ruby container.

**Comment by paneq** [comment]
> I am sure you will agree that finding one timing difference does not imply that there is a computational overhead.

I agree, that's why I mentioned the issues I experience with some other workloads. But this one was the easiest to reproduce.

> Would you try the following?

That's a lovely counter-example indeed on my machine:

```
time docker run ruby:2.6.5 ruby -e "(1..100_000).inject(:*)"
        4.60 real         0.04 user         0.02 sys
time docker run ruby:2.6.5 ruby -e "(1..100_000).inject(:*)"
        3.80 real         0.03 user         0.01 sys
time docker run ruby:2.6.5 ruby -e "(1..100_000).inject(:*)"
        3.73 real         0.03 user         0.01 sys

time ruby -e "(1..100_000).inject(:*)"
        7.39 real         4.14 user         2.75 sys
time ruby -e "(1..100_000).inject(:*)"
        7.17 real         4.15 user         2.67 sys
time ruby -e "(1..100_000).inject(:*)"
        7.12 real         4.13 user         2.58 sys
```

Now I am even more baffled... I will try to find some open-source test suites which we can try executing for more data points. Thank you for looking into it.

**Comment by paneq** [comment]
##### Setup

* https://github.com/paneq/rails/pull/1/files
* `docker-compose up`
* `docker-compose exec devrails bash`
* Every benchmark was run multiple times to ensure the results are consistent

##### Thoughts

* It might be that some file operations are slow and I am just not aware of how many of them these tests perform.
* There are plenty of info around about osxfs and perf issues for mounted volumes. That's why I perform these tests without mounting anything and with the whole application/benchmark code inside the docker image.

##### Rails actionpack test:

```
cd actionpack
time bundle exec rake test
```

* Native

```
Finished in 12.166587s, 274.9333 runs/s, 1336.9403 assertions/s.
3345 runs, 16266 assertions, 0 failures, 0 errors, 0 skips

       17.09 real        39.65 user         3.15 sys
```

* Docker

```
Finished in 48.023361s, 69.6328 runs/s, 338.6477 assertions/s.
3344 runs, 16263 assertions, 0 failures, 0 errors, 0 skips

real	0m50.853s
user	0m44.688s
sys	0m2.303s
```

##### Rails actionview test:

```
cd actionview
time bundle exec rake test
```

* native

```
       13.05 real        14.05 user         3.75 sys
```

* docker

```
real	0m17.547s
user	0m13.378s
sys	0m2.107s
```

##### Rails activemodel test:

```
cd activemodel
time bundle exec rake test
```

* native

```
Finished in 1.755843s, 519.9782 runs/s, 1446.5986 assertions/s.
913 runs, 2540 assertions, 0 failures, 0 errors, 0 skips

        3.44 real         2.94 user         1.14 sys
```

* docker

```
Finished in 2.891900s, 315.7094 runs/s, 878.3152 assertions/s.
913 runs, 2540 assertions, 0 failures, 0 errors, 0 skips

real	0m4.522s
user	0m3.030s
sys	0m0.992s
```

**Comment by paneq** [comment]
##### Setup

* https://github.com/paneq/material-ui/pull/1/files
* `docker-compose build`
* `docker-compose up`
* `docker-compose exec devmaterialui bash`
* `time yarn test:unit`
* The benchmark was run multiple times to ensure the results are consistent

##### Results

* Native

```
  2974 passing (28s)
  14 pending

✨  Done in 37.14s.

       37.69 real        36.72 user         3.04 sys
```

* Docker

```
  2974 passing (1m)
  14 pending

Done in 71.96s.

real	1m12.542s
user	0m58.989s
sys	0m6.515s
```

**Comment by dennissivia** [comment]
Hey 👋 ,
thanks for taking a look into this.
There are a few things I would like to add.

1. Make sure you perform "warm-up" runs for your benchmark. Not only does this deal with all kinds of RAM/Disk caching factors but also makes sure your CPU runs a maximum performance and the first run does not start in with some powersave based CPU frequency.

1. I ran the above example with 2 warups und 5 rounds (which seems to be fine, given the σ, but still not a lot). My conclusion with the `inject` test, as well as the initial `addition` test is, that both version run in similar time if the test-period is small (subsecond or few seconds). The longer the workload runs, the more the difference is visiable.

Example from the `inject`-based test:

100.000 
```
$ hyperfine --min-runs 5 --warmup 2 'ruby -e "(1..100_000).inject(:*)"'
Benchmark #1: ruby -e "(1..100_000).inject(:*)"
  Time (mean ± σ):      2.702 s ±  0.061 s    [User: 2.383 s, System: 0.310 s]
  Range (min … max):    2.626 s …  2.784 s    5 runs

$hyperfine --min-runs 5 --warmup 2 'docker run ruby:2.6.3 ruby -e "(1..100_000).inject(:*)"'    
Benchmark #1: docker run ruby:2.6.3 ruby -e "(1..100_000).inject(:*)"
  Time (mean ± σ):      2.990 s ±  0.042 s    [User: 26.9 ms, System: 15.6 ms]
  Range (min … max):    2.941 s …  3.038 s    5 runs
```


300.000
```
$ hyperfine --min-runs 5 --warmup 2 'ruby -e "(1..300_000).inject(:*)"'                     
Benchmark #1: ruby -e "(1..300_000).inject(:*)"
  Time (mean ± σ):     21.523 s ±  0.086 s    [User: 20.951 s, System: 0.536 s]
  Range (min … max):   21.391 s … 21.603 s    5 runs

$hyperfine --min-runs 5 --warmup 2 'docker run ruby:2.6.3 ruby -e "(1..300_000).inject(:*)"'     
Benchmark #1: docker run ruby:2.6.3 ruby -e "(1..300_000).inject(:*)"
  Time (mean ± σ):     25.563 s ±  0.387 s    [User: 27.5 ms, System: 15.5 ms]
  Range (min … max):   24.977 s … 26.020 s    5 runs
```

The same was observable with the addition based test. with small values the time was similar, for larger values, the gab was already significant. So we should make sure we pick workloads that are not too small for the discussion.
That's why I think it made sense that @paneq used real world projects https://github.com/docker/for-mac/issues/4769#issuecomment-660623338 and https://github.com/docker/for-mac/issues/4769#issuecomment-660623338 which confirmed the difference as well.

**Comment by lemire** [comment]
My conclusion so far is that the performance of ruby on macos cannot be directly compared with the performance ruby in from the ruby image. Even if the ruby version is the same, it does not make an equivalent system. According to my measures, I am getting that on one test (the computation of the factorial), running ruby in a docker image boosts the performance by 20%. This cannot be because docker is able to run computational tasks faster than native. It has to be because the binary (ruby) and its sub-libraries are different.


For example, do you know that the ruby interpreter was compiled against the same compiler and using the same standard libraries? Recompiling a system like Ruby using a different compiler can certainly, on some tasks, explain a 10%, 20% difference.

The binaries are not even approximately the same size:


<img width="599" alt="Screen Shot 2020-07-20 at 10 03 42 AM" src="https://user-images.githubusercontent.com/391987/87946652-4cb06d00-ca70-11ea-9066-5e79ce0ff62e.png">


So I'd like to read an argument that would reassure me that we should expect identical performance from the two ruby systems.

(This is quite aside the fact that docker has well documented overhead when it comes to network and disk access, and any other syscall... though I think we agree that syscalls are outside our realm of discussion here.)

@dennissivia My results do not agree with yours: they are in reverse. I get that ruby on docker is consistently faster than ruby on mac... 

<img width="985" alt="Screen Shot 2020-07-20 at 9 21 09 AM" src="https://user-images.githubusercontent.com/391987/87942273-5c2cb780-ca6a-11ea-878c-b30c8927954c.png">


Let us estimate the docker overhead...

<img width="800" alt="Screen Shot 2020-07-20 at 9 26 35 AM" src="https://user-images.githubusercontent.com/391987/87942793-1cb29b00-ca6b-11ea-896a-513947410245.png">

If you take the docker overhead into account (0.4s), you consistently get that the ruby-on-docker version is faster by about 20%.


Here is my ruby version...

<img width="623" alt="Screen Shot 2020-07-20 at 9 24 02 AM" src="https://user-images.githubusercontent.com/391987/87942560-c2193f00-ca6a-11ea-8410-5d340b2d543b.png">

Here is my system:

<img width="1062" alt="Screen Shot 2020-07-20 at 9 24 24 AM" src="https://user-images.githubusercontent.com/391987/87942593-cfcec480-ca6a-11ea-9044-7288202cd8cd.png">

```
docker --version
Docker version 19.03.8
```


> Make sure you perform "warm-up" runs for your benchmark. Not only does this deal with all kinds of RAM/Disk caching factors but also makes sure your CPU runs a maximum performance and the first run does not start in with some powersave based CPU frequency.

On a good machine, the processor can switch frequency in microseconds (10 us), so compared with a 3 second benchmark, the frequency shifting cost is going to be negligible. But of course, you should not time the first call to the program (docker/ruby) when it is cold, since hitting the disk is going to be expensive.

It seems that @paneq is getting a lot of run-to-run variation. I do not. My numbers are flat. Of course, I am using a well ventilated desktop. Never benchmark on a laptop.


> My conclusion with the inject test, as well as the initial addition test is, that both version run in similar time if the test-period is small (subsecond or few seconds). The longer the workload runs, the more the difference is visiable.


How do you explain this effect?

I have a nice explanation for my result. There is an overhead due to launching the docker container of about 0.4 s. If I deduct this overhead then I get a flat 80% difference.

That is, from first principles, you assume that short tests penalize the docker runs by a fixed amount of time, due to the overhead of launching the docker container. 

> So we should make sure we pick workloads that are not too small for the discussion.

As long as the docker overhead is fixed, then I think we do not need to go to very long durations. I would argue that 3 seconds is already more than long enough. Of course, if you want to make the cost of launching the docker container negligible, then you need much longer than 3 s, but if you can count on it being a constant, then you can just mentally deduct it.

**Comment by lemire** [comment]
It might be interesting to profile the Ruby code...

https://github.com/rbspy/rbspy

@jvns

**Comment by paneq** [comment]
It turned out that rails tests (actionpack, etc) are multi-threaded and not all CPUs were assigned to my docker instance. I apologize for presenting misleading data. Benchmarking is hard.

When running:

```
MT_CPU=1 time -p bundle exec rake test
```

The results are:

* native

```
Finished in 20.727217s, 161.3820 runs/s, 784.4758 assertions/s.
3345 runs, 16260 assertions, 0 failures, 0 errors, 0 skips
real        25.83
user        22.40
sys          1.87
```

* docker

```
Finished in 37.185518s, 89.9275 runs/s, 437.1863 assertions/s.
3344 runs, 16257 assertions, 0 failures, 0 errors, 0 skips
real 40.04
user 38.68
sys 0.98
```

> My conclusion so far is that the performance of ruby on macos cannot be directly compared with the performance ruby in from the ruby image. Even if the ruby version is the same, it does not make an equivalent system.
> ...
> So I'd like to read an argument that would reassure me that we should expect identical performance from the two ruby systems.

I see your point but I am not sure what would be the sufficient reproduction threshold that would be convincing. Can we establish that? Would having the numbers from same machine running docker on linux be beneficial?

> (This is quite aside the fact that docker has well documented overhead when it comes to network and disk access, and any other syscall... though I think we agree that syscalls are outside our realm of discussion here.)

* Where can I read about those?
* Is there an overhead for disk access to files in the docker image (not mounted)?
* I believe syscalls are outside of the realm but I wouldn't exclude anything.

The only thing that I found consistently (and also officially documented) in the Internets are about [mounted volumes](https://docs.docker.com/docker-for-mac/osxfs-caching/).

In the meantime, I will try to see what rbspy can tell me.

**Comment by lemire** [comment]
> I see your point but I am not sure what would be the sufficient reproduction threshold that would be convincing. Can we establish that? Would having the numbers from same machine running docker on linux be beneficial?

Using different ruby interpreters, it can be difficult, I fear, to establish that there is a sizeable overhead on purely computational tasks.

So a fairer statement is that, in practice, you find that the native macos ruby interpreter has much better performance than the docker-mounted ruby interpreter. But if you don't know that these two interpreter are computationally equivalent, it is hard to say more.

With a proper profiler, it should be possible to tell where the time goes. 

> Is there an overhead for disk access to files in the docker image (not mounted)?

It stands to reason that there would be an overhead, right? Your program must talk to the "system" which is docker and its VM, then docker and VM must talk to the host (in general). And then it comes back.

If, ultimately, your files are written down on your disk, then they have to go through docker and then to the host.

**Comment by paneq** [comment]
One more data point from one developer. Very similar numbers, even slightly faster in docker, but a lot variability.

```
a@MacBook-Pro:~/Projects/my/mobx|mobx6
⇒  hyperfine --min-runs 5 --warmup 2 'yarn test'
Benchmark #1: yarn test
  Time (mean ± σ):     13.710 s ±  2.419 s    [User: 81.395 s, System: 21.956 s]
  Range (min … max):   10.587 s … 16.070 s    5 runs

a@MacBook-Pro:~/Projects/my/mobx|mobx6
⇒  hyperfine --min-runs 5 --warmup 2 'docker exec mobx_devmaterialui_1 yarn test'
Benchmark #1: docker exec mobx_devmaterialui_1 yarn test
  Time (mean ± σ):     13.111 s ±  1.869 s    [User: 36.1 ms, System: 17.6 ms]
  Range (min … max):   11.796 s … 16.278 s    5 runs
```

I am looking more deeply into some of the tests which we used and I have some hypothesis but I will post later when I verify it more.

**Comment by dennissivia** [comment]
Hi again 👋 ,
I took another look at the problem and I have a few updates.
First: I created a repository for all the scripts I am using, so that it can be reproduced on other people's system as well as on servers. (which is what I did to have a comparison with a Linux based host system). Everything is available [here](https://github.com/dennissivia/docker-mac-performance).

##  Behavior for non Ruby cases

To make it easier to track down the problem I wrote a small c(++) Program that runs a loop and does some arithmetic and logical operations. I tried to make them hard to optimize and also compiled the program with `-O0` to skip optimizations.
The [source](https://github.com/dennissivia/docker-mac-performance/blob/main/docker-bench.cpp) is available in the same repository.

### Results
<img width="777" alt="Screenshot 2020-07-30 at 17 57 41" src="https://user-images.githubusercontent.com/3408791/88945337-2c618900-d28e-11ea-97fc-09751f031a42.png">

The outcome of running the C++ program is, that there is no overhead for when it runs in docker compared to the host system.
This confirms earlier results, when I performed an initial test with `sysbench` which yielded identical results on the host and the container system.

## Ruby

### GC

We used two different approaches in this thread. @paneq suggested `300_000_000.times{1+1}` in his [here](https://github.com/docker/for-mac/issues/4769#issue-655235213), while in this reply the following method was suggested: 
`(1..100_000).inject(:*)`.

The later has two important differences: 
* It is hard to optimize away/inline by the runtime/JIT, since it does not use addition of literals. Though this makes it more realistic, it also means there are object allocations in a loop. 
This creates a lot of GC pressure. I measured the GC impact with [this script](https://github.com/dennissivia/docker-mac-performance/blob/main/ruby-bench-gc-impact.rb)
During the run of the loop there are ~2700 GC runs that cost ~0.5seconds (on my system), but also cause a lot of unpredictability / variance to the benchmark.

#### Conclusion

I updated the scripts to run `GC.disable` first and suggest that we all stick with `300_000_000.times{1+1}` for simple CPU-bound ruby benchmarks. Or any other command, that does not run  into memory issues (The  above "inject" required >6GB  without GC). 


### Normalization

In order to ensure that the results are as comparable as possible (given that we have 2 different operating systems with different compiler and package systems), I decided to compile Ruby from source using `clang`. I could not use identical versions since custom Apple patches are used on MacOS. However, I used versions that are as close as possible and tested the impact on Linux (server) and MacOS (My MacBook Pro)

### Linux 
<img width="780" alt="Screenshot 2020-07-30 at 14 31 12" src="https://user-images.githubusercontent.com/3408791/88945974-ece76c80-d28e-11ea-8948-4620a4958532.png">

One observation is that there is little to no impact on Linux. Container and Host system perform with almost identical time behavior.
This is true for the bundled ruby as well as the self-compiled ruby (2.5.5) 
One thing that is  worth mentioning, is that using debian (on the host or in a container) and it's re-build package is always faster than using a self-compiled Ruby or any of the ruby-specific docker images. 
I would explain this with compiler optimizations for the release-package or something along those lines. This is not important to this thread but might be worth checking in different contexts. `cc` @paneq 

### MacOS

<img width="1097" alt="Screenshot 2020-07-30 at 17 37 32" src="https://user-images.githubusercontent.com/3408791/88945999-f7096b00-d28e-11ea-90e8-96e506a3cc5d.png">

On MacOS the picture is much different. All 3 native runs perform in more or less the same time.
So independent of the compilation (self compiled, vs packaged) and the version (2.5.5 vs 2.6.5), the time to run the loop is always below 10 seconds.

For the docker run, there is an important difference: 
**Ruby 2.6.x performs much slower than Ruby 2.5.5**. This is independent of the compilation method and even disabling the Ruby JIT (which was introduced in 2.6) did not make a difference. I also used a tempfs mount, since the JIT writes to `/tmp` before compiling the file. (See [this link](https://bugs.ruby-lang.org/projects/ruby/wiki/MJIT) and other sources for details)


### Conclusion

* Independent of the way we build (compile) Ruby, our example loop does not yield a difference when we run it on the host- or the container on Linux based systems
* On MacOS there is a time behavior of 20% or more depending on the Ruby version. This is confirmed by running a manually compiled Ruby 2.5.5 version on the host and in a container.  (clang has been used to compile it)
* There is a significant performance difference between 2.6.x and 2.5x which might be related to the JIT. This is not relevant for this issue, since we can focus on 2.5.5 to simplify the result
* Debian + ruby package are significantly faster on host and container, than official Ruby images

## Next steps

Next we should run a profiler when we run Ruby in a container to see if there are any IO waits, which are a strong candidate, given that the C++ and `sysbench` based tests did not yield a difference in time behavior.

**Comment by paneq** [comment]
#### Intro

Nice job @dennissivia . Here is what I've done in the meantime.

I compared the speed of rails actionpack tests between docker and native run to see if there are some outliers and there were:

```
TestRoutingMapper
6.785

RequestForgeryProtectionControllerUsingResetSessionTest
3.06

RequestForgeryProtectionControllerUsingExceptionTest
3.05

YieldingAroundFiltersTest
0.581

ActionPack::URLForIntegrationTest
0.56

ResourcesTest
0.414

RouteSetTest
0.358

RoutingAssertionsTest
0.226
```

so I decided to focus on `TestRoutingMapper` (`test/dispatch/routing_test.rb`). There is `test_shallow_nested_resources` case inside which does 25 rails requests. Running just that is enough to reproduce. I added some logging to see how long particular rails requests take in that test.

#### native

```
~/d/t/r/r/actionpack (dockerized|✚3…) $ env MT_CPU=0 bundle exec ruby -Ilib:test test/dispatch/routing_test.rb --name test_shallow_nested_resources --seed 62879
Run options: --name test_shallow_nested_resources --seed 62879

# Running:

SETUP  : 1596180773.3300 : 0.0000
START  : 1596180773.3300 : 0.0001
API1001: 1596180773.3395 : 0.0094
API1002: 1596180773.3584 : 0.0190
API1003: 1596180773.3660 : 0.0075
API1004: 1596180773.3728 : 0.0068
API1005: 1596180773.3826 : 0.0098
API1006: 1596180773.3947 : 0.0121
API1007: 1596180773.4078 : 0.0131
API1008: 1596180773.4201 : 0.0123
API1009: 1596180773.4366 : 0.0165
API1010: 1596180773.4442 : 0.0076
API1011: 1596180773.4521 : 0.0079
API1012: 1596180773.4601 : 0.0080
API1013: 1596180773.4665 : 0.0064
API1014: 1596180773.4728 : 0.0063
API1015: 1596180773.4797 : 0.0069
API1016: 1596180773.4860 : 0.0064
API1017: 1596180773.4931 : 0.0071
API1018: 1596180773.5002 : 0.0070
API1019: 1596180773.5070 : 0.0068
API1020: 1596180773.5147 : 0.0078
API1021: 1596180773.5212 : 0.0065
API1022: 1596180773.5216 : 0.0004
API1023: 1596180773.5219 : 0.0003
API1024: 1596180773.5222 : 0.0003
API1025: 1596180773.5225 : 0.0003
FINISHED: 1596180773.5297 : 0.0072
TEARDOWN: 1596180773.5298 : 0.0001
.

Finished in 0.211749s, 4.7226 runs/s, 236.1286 assertions/s.
1 runs, 50 assertions, 0 failures, 0 errors, 0 skips
```


#### docker

```
root@19836e1db75e:/rails/actionpack# env MT_CPU=0 bundle exec ruby -Ilib:test test/dispatch/routing_test.rb --name test_shallow_nested_resources --seed 62879
Run options: --name test_shallow_nested_resources --seed 62879

# Running:

SETUP  : 1596180764.9738 : 0.0000
START  : 1596180764.9739 : 0.0001
API1001: 1596180764.9878 : 0.0139
API1002: 1596180765.0082 : 0.0204
API1003: 1596180765.0229 : 0.0146
API1004: 1596180765.0370 : 0.0141
API1005: 1596180765.0522 : 0.0152
API1006: 1596180765.0670 : 0.0148
API1007: 1596180765.0826 : 0.0156
API1008: 1596180765.0981 : 0.0155
API1009: 1596180765.1155 : 0.0174
API1010: 1596180765.1318 : 0.0163
API1011: 1596180765.1475 : 0.0157
API1012: 1596180765.1682 : 0.0207
API1013: 1596180765.1846 : 0.0164
API1014: 1596180765.1990 : 0.0145
API1015: 1596180765.2155 : 0.0164
API1016: 1596180765.2317 : 0.0162
API1017: 1596180765.2456 : 0.0139
API1018: 1596180765.2608 : 0.0151
API1019: 1596180765.2766 : 0.0158
API1020: 1596180765.2919 : 0.0153
API1021: 1596180765.3069 : 0.0150
API1022: 1596180765.3077 : 0.0008
API1023: 1596180765.3085 : 0.0007
API1024: 1596180765.3091 : 0.0006
API1025: 1596180765.3097 : 0.0006
FINISHED: 1596180765.3147 : 0.0050
TEARDOWN: 1596180765.3148 : 0.0001
.

Finished in 0.351205s, 2.8473 runs/s, 142.3671 assertions/s.
1 runs, 50 assertions, 0 failures, 0 errors, 0 skips
```

#### Docker strace

My hypothesis was that there is some IO operation performed between those rails requests so I used `strace` to check it.
You can find it here: https://gist.github.com/paneq/b4da624f54564d6b2ae5c79bcfe9c811
There are some at the beginning and after the test is finished but not during most of them. And yet almost every request within that test is slower. Which suggests it is not an IO issue.

#### Profiling

I tried 3 profiles:

* rbspy - I could not run it within docker
* stackprof - flamegraphs looked identical and I could not find any meaningful difference
* ruby-prof - profiling within docker took 1 or 2 orders of magnitude longer than outside docker, comparing these numbers together did not make sense


#### Summary

I still know nothing.

**Comment by lemire** [comment]
@paneq @dennissivia 

I think that what "300_000_000.times{1+1}" does is to effectively call a function millions of times, with the accompanying cost of setting up the stack, updating the call stack, and so forth.

I don't think it is likely to generate garbage (it is all on stack), or IO of any kind.

**Comment by dennissivia** [comment]
> I don't think it is likely to generate garbage (it is all on stack), or IO of any kind.

I agree that it does not create garbage. That's what I was stating above. Addition of literals is optimized by the Ruby VM. However, the inject loop, does create garbage since any Number in Ruby is an object (unless optimized away). So given 2 numbers `foo` and `bar` then `baz = foo * bar` is managing 3 objects on the heap which are garbage collected later. That's what I showed with the above example about GC impact of `1+1` vs `inject(&:*)`. 
One thing to mention is that the Ruby 2.6 MJIT is especially trying to optimize addition and loops. Since the (current simplistic) JIT implementation in Ruby does write a C file to Disk (/tmp) and compiles that with LLVM, disk IO happens with the JIT is used. 
This means there is IO, in case of the JIT and there is GC in case of non trivial arithmetic operations, since Numbers are Heap allocated objects in Ruby.

That all said, the simple `300_000_000.times{1+1}` is  not triggering a GC, which I confirmed with the `GC::Profiler` (see script link above) and Ruby 2.5.5 does not have a JIT. So to me it is unclear, why there is a significant difference of >20% in time behavior, even after excluding these factors.
